### ISSUE TITLE
Couchbase Server 4.1 Error Executing Query With Prefiltered All Subquery

### ISSUE THREAD

**Issue by brantburnett** [couchbaselabs/Linq2Couchbase#89]
The integration test AnyAllTests_AllNestedArrayPrefiltered is failing.  The generated N1QL query is:

```
SELECT `Extent1`.`name` as `name`, `Extent1`.`address` as `address`
FROM `beer-sample` as `Extent1`
WHERE ((`Extent1`.`type` = 'brewery')
AND EVERY `Extent2` IN (ARRAY `Extent2` FOR `Extent2` IN `Extent1`.`address` WHEN (`Extent2` = '563 Second Street') END) SATISFIES (`Extent2` = '101 Fake Street') END)
ORDER BY `Extent1`.`name` ASC
```

This test was not failing against Couchbase Server 4.0.  The reuse of Extent2 outside the ARRAY subquery for the EVERY statement is apparently not allowed by the the Couchbase Server 4.1 query engine.

**Comment by jeffrymorris** [comment]
@brantburnett -

I opened a server ticket for an explanation on this: https://issues.couchbase.com/browse/MB-17170

-Jeff

**Comment by jeffrymorris** [comment]
@brantburnett -

It looks like EVERY and ANY should introduce a new variable "Extent3" in the query:

{code}
SELECT `Extent1`.`name` as `name`, `Extent1`.`address` as `address`
 FROM `beer-sample` as `Extent1`
 WHERE ((`Extent1`.`type` = 'brewery')
 AND EVERY `Extent3` IN (ARRAY `Extent2` FOR `Extent2` IN `Extent1`.`address` WHEN (`Extent2` = '563 Second Street') END) SATISFIES (`Extent3` = '101 Fake Street') END)
 ORDER BY `Extent1`.`name` ASC
{code}

That's the official word from the query team.

-Jeff

**Comment by brantburnett** [comment]
@jeffrymorris 

Yeah, I thought they would say that.  I've just played with it a little so far, but it looks like now you can reference identifiers from outside the ARRAY statement inside the ARRAY statement.  Thus they can't allow identifier conflicts anymore.  It's going to be tricky to fix in LINQ, though.

Brant

**Comment by brantburnett** [comment]
Fixed, see 48a80fa
